File: jsfring-webservice/src/main/java/fr/pinguet62/jsfring/webservice/controller/UserController.java
package fr.pinguet62.jsfring.webservice.controller;

import fr.pinguet62.jsfring.model.sql.User;
import fr.pinguet62.jsfring.service.UserService;
import fr.pinguet62.jsfring.webservice.converter.UserMapper;
import fr.pinguet62.jsfring.webservice.dto.UserDto;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import static fr.pinguet62.jsfring.webservice.controller.UserController.PATH;
import static java.util.Objects.requireNonNull;

@RestController
@RequestMapping(PATH)
public final class UserController {

    public static final String PATH = "/user";

    private final UserService userService;

    private final UserMapper converter;

    public UserController(UserService userService, UserMapper converter) {
        this.userService = requireNonNull(userService);
        this.converter = requireNonNull(converter);
    }

    @PutMapping
    public void create(@RequestBody UserDto userDto) {
        User user = converter.fromDto(userDto);
        userService.create(user);
    }

    @GetMapping("/{email:.+}")
    public Mono<UserDto> get(@PathVariable String email) {
        return userService
                .get(email)
                .map(converter::toDto);
    }

    @GetMapping
    public Flux<UserDto> list() {
        return userService
                .getAll()
                .map(converter::toDto);
    }

    @PostMapping
    public void update(@RequestBody UserDto userDto) {
        userService
                .get(userDto.getEmail())
                .doOnNext(user -> converter.updateFromDto(user, userDto))
                .doOnNext(userService::update);
    }

}

File: jsfring-webservice/src/main/kotlin/fr/pinguet62/jsfring/webservice/controller/UserController.kt
package fr.pinguet62.jsfring.webservice.controller

import fr.pinguet62.jsfring.model.sql.User
import fr.pinguet62.jsfring.service.UserService
import fr.pinguet62.jsfring.webservice.controller.UserController.Companion.PATH
import fr.pinguet62.jsfring.webservice.converter.UserMapper
import fr.pinguet62.jsfring.webservice.dto.UserDto
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Flux

@RestController
@RequestMapping(PATH)
class UserController(
        val userService: UserService,
        val converter: UserMapper
) {

    companion object {
        const val PATH = "/user"
    }

    @PutMapping
    fun create(@RequestBody userDto: UserDto) {
        var user: User = converter.fromDto(userDto)
        userService.create(user)
    }

    @GetMapping("/{email:.+}")
    fun get(@PathVariable email: String) =
            userService
                    .get(email)
                    .map { converter.toDto(it) }

    @GetMapping
    fun list(): Flux<UserDto> =
            userService
                    .getAll()
                    .map { converter.toDto(it) }

    @PostMapping
    fun update(@RequestBody userDto: UserDto) {
        userService
                .get(userDto.email)
                .doOnNext { converter.updateFromDto(it, userDto) }
                .doOnNext { userService.update(it) }
    }

}